File: cs/Platforms/Client.Wpf/ComponentModel/ViewModel.cs
namespace Client.Wpf
{
    using System.ComponentModel;

    public abstract class ViewModel : INotifyPropertyChanged, INotifyPropertyChanging
    {
        #region ---------- Protected constructors ----------

        protected ViewModel()
            : base()
        {
        }

        #endregion

        #region ---------- Public events ----------

        public event PropertyChangedEventHandler PropertyChanged;
        public event PropertyChangingEventHandler PropertyChanging;

        #endregion

        #region ---------- Protected methods ----------

        protected virtual void OnPropertyChanged(PropertyChangedEventArgs e)
        {
            this.PropertyChanged?.Invoke(this, e);
        }

        protected virtual void OnPropertyChanging(PropertyChangingEventArgs e)
        {
            this.PropertyChanging?.Invoke(this, e);
        }

        protected void RaisePropertyChanged(string propertyName)
        {
            this.OnPropertyChanged(new PropertyChangedEventArgs(propertyName));
        }

        protected void RaisePropertyChanging(string propertyName)
        {
            this.OnPropertyChanging(new PropertyChangingEventArgs(propertyName));
        }

        #endregion
    }
}

File: vb/Platforms/Client.Wpf/ComponentModel/ViewModel.vb
Imports System.ComponentModel

Namespace Client.Wpf
    Public MustInherit Class ViewModel
        Implements INotifyPropertyChanged, INotifyPropertyChanging

        '---------- Protected constructors ----------

        Protected Sub New()
            MyBase.New()
        End Sub

        '---------- Public events ----------

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
        Public Event PropertyChanging As PropertyChangingEventHandler Implements INotifyPropertyChanging.PropertyChanging

        '---------- Protected methods ----------

        Protected Overridable Sub OnPropertyChanged(ByVal e As PropertyChangedEventArgs)
            RaiseEvent PropertyChanged(Me, e)
        End Sub

        Protected Overridable Sub OnPropertyChanging(ByVal e As PropertyChangingEventArgs)
            RaiseEvent PropertyChanging(Me, e)
        End Sub

        Protected Sub RaisePropertyChanged(ByVal propertyName As String)
            Me.OnPropertyChanged(New PropertyChangedEventArgs(propertyName))
        End Sub

        Protected Sub RaisePropertyChanging(ByVal propertyName As String)
            Me.OnPropertyChanging(New PropertyChangingEventArgs(propertyName))
        End Sub
    End Class
End Namespace
